### ISSUE TITLE
Type-narrowing `Type[Any]` by comparison

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#5447]
#### Describe the bug

Adding a check to a type reference (e.g. equality or membership) doesn't narrow the possible type, and with an optional type it remains optional even if all the possibilities are not `None`.

#### To reproduce

Minimal repro:

```python
from typing import Any, Optional, Type

def test(dtype: Optional[Type[Any]]):
    if dtype == int:
        reveal_type(dtype)  # Type of "dtype" is "type | None"
```

Slightly fuller example of what I'm actually doing, handling a couple of primitive types directly:

```python
class Field(Generic[_TAny]):
    data_type: Optional[Type[_TAny]] = None

    def encode(self, value: Any) -> Any:
        if self.data_type in (int, float, bool):
            return self.data_type(value)  # Object of type "None" cannot be called
        else:
            return str(value)

class IntField(Field[int]):
    data_type = int
```

#### Expected behavior

Ideally, `dtype` in the first example to be `type[int]`, or at least `type[Any]` rather than `type[Any] | None`.

Of course I can add a truthy check on `dtype` to get around it, but this seems like an inference that should be made already?  There's also `issubclass` which type narrows correctly, but then requires explicit handling of the `None` case anyway.

#### VS Code extension or command-line

VSCode: Pylance v2023.7.10 / Pyright v1.1.316

**Comment by cainmagi** [comment]
I need to tell you that the coding type you are using for validating type is incorrect. The correct way is:

```python
import inspect

if inspect.isclass(dtype) and issubclass(dtype, (int, float, bool)):
    # Do something here
    # You will find that dtype has been annotated as `type[int] | type[float] | type[bool]` now
    ...
```

Here `inspect.isclass` ensures that `dtype` is a subclass of `type`. `issubclass` provides the type check and requires the first argument being a subclass of `type`.

Certainly, if you would like, you can write a customized `TypeGuard` function to do that:

```python
import inspect
from typing import Union, Any
from typing_extensions import TypeGuard


def is_numeric(val: Any) -> TypeGuard[Union[type[int], type[float], type[bool]]]:
    return inspect.isclass(val) and issubclass(val, (int, float, bool))


if isnumeric(dtype):
    # Do something here
    # You will find that dtype has been annotated as `type[int] | type[float] | type[bool]` now
    ...
```

**Comment by Terrance** [comment]
As I said above, I could do the `issubclass` check -- I'm mainly interested in the (lack of) narrowing to what is effectively a literal but of a `type` object.

Looks like I can reproduce this with an `object` reference too (or a reference to a general class instance), rather than a type:

```python
sentinel = object()

def test(value: Optional[object]):
    if value == sentinel:
        reveal_type(value)  # Type of "value" is "object | None"
```

**Comment by cainmagi** [comment]
> As I said above, I could do the `issubclass` check -- I'm mainly interested in the (lack of) narrowing to what is effectively a literal but of a `type` object.
> 
> Looks like I can reproduce this with an `object` reference too (or a reference to a general class instance), rather than a type:
> 
> ```python
> sentinel = object()
> 
> def test(value: Optional[object]):
>     if value is sentinel:
>         reveal_type(value)  # Type of "value" is "object | None"
> ```

To my understanding, types like `int`, `bool`, or something **is not "literal"**, otherwise they can be used with `Literal`. For example, the following codes will raise type check failure:
```python
from typing_extensions import Literal

Literal[int, bool]  # Cannot let type be the member of Literal.
```

For example, we can derive the `int` or `bool` like this:

```python
from typing import NewType

class NewInt(int):
   ...

NewInt2 = NewType('NewInt2', int)

def func(val: type[int]):
    ...
```

In the above example, we can pass classes like `NewInt` to `func()`, which means that `val` is possibly to be the subtype of `int`, thus using `val is int` will not "narrow down" this type.

-------

The second example you provide can justify this issue more precisely:

```python
sentinel = object()

another = object()

def test(value: Optional[object]):
    if value is sentinel:
        reveal_type(value)  # Type of "value" is "object | None"
    else:
        reveal_type(value)  # If value is narrowed down, this branch should return `None`
```

In the above example, both `sentinel` and `another` are `object()`, which means they can be passed to `test()`. However, using `value is sentinel` cannot guarantee that `value is not another`. In the above example, if `value` is "narrowed down", the else branch should be able to be narrowed to `None`. Apparently, `value is sentinel` cannot provide such confidence.

**Comment by erictraut** [comment]
@Terrance, this is not a supported type guard in pyright. See [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) for a complete list.

You're asking for two new type guard patterns: `x == <class>` and `x in <tuple of classes>`. We do occasionally add support for new patterns, but the bar is relatively high because each pattern requires significant custom logic to handle both the positive ("if") and negative ("else") narrowing cases. They also potentially slow down type analysis. For example, the `==` pattern that you're requesting would cause significant slow-downs, so it's unlikely that we'd add it. The reason it affects performance is that all conditional expressions of the form `x == <expr>` and `x != <expr>` would need to analyze the type of `<expr>` to determine whether it resolves to a class. This would create many false dependencies between `x` (the symbol potentially being narrowed) and any other symbols in `<expr>`. These false dependencies would make it harder for types to converge during analysis, especially when loops are involved. 

As others have pointed out, `x == <class>` is problematic for other reasons. The `==` and `!=` operators can be overloaded in Python, so it's generally unsafe for a static type checker to make assumptions about their behavior.

You're the first person (as far as I can remember) who has requested the `x == <class>` and `x in <tuple of classes>` patterns. I think someone has requested the `x is <class>` pattern previously for use with sentinels. If that's something you're interested in, please search for that issue and upvote it. That pattern would be more feasible to support (from a perf perspective) because the `is` operator is much less common than `==` or `!=`.

**Comment by Terrance** [comment]
I agree `==` isn't the best check, but for multiple possible values there isn't a particularly nice way to do identity checks (AIUI `__contains__` in the built-in containers use `==` rather than `is` for membership tests).

`is` would be a safer expression to support, though from some quick searches I can't find an obvious previous issue for it, perhaps because "is" is hard to search for!

---

> See [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) for a complete list.

For what it's worth, `x in y` is listed as a type guard there, but the tuple example I had above doesn't do that:

> `x in y` or `x not in y` (where y is instance of list, set, frozenset, deque, tuple, dict, defaultdict, or OrderedDict)

I assume this is only valid if `y` contains items of types described by the various `==` guards further up (i.e. literals, `None`, `Ellipsis` etc.)?
